p2p-sync/src/main/java/org/smartregister/p2p/presenter/P2PReceiverPresenter.java

```diff
@@ -165,65 +172,256 @@ public void onAuthorizationDetailsProvided(@NonNull Map<String, Object> authoriz
+            }
+        }
+    }
+
+    @Override
+    public void processHashKey(@NonNull final String endpointId, @NonNull Payload payload) {
+        if (payload.getType() == Payload.Type.BYTES && payload.asBytes() != null) {
+            String payloadAsString = new String(payload.asBytes());
+
+            Map<String, Object> basicDeviceDetails = null;
+            try {
+                basicDeviceDetails = (Map<String, Object>) new Gson()
+                        .fromJson(payloadAsString, Map.class);
+            } catch (JsonSyntaxException e) {
+                Timber.e(e);
+            }
+
+            if (basicDeviceDetails == null || basicDeviceDetails.get(Constants.BasicDeviceDetails.KEY_DEVICE_ID) == null) {
+                Timber.e(view.getString(R.string.log_hash_key_sent_was_null));
+                disconnectAndReset(endpointId);
             } else {
-                // Should get the details to authorize
-                if (payload.getType() == Payload.Type.BYTES && payload.asBytes() != null) {
-                    String authenticationDetailsJson = new String(payload.asBytes());
+                connectionLevel = ConnectionLevel.RECEIVED_HASH_KEY;
+
+                // Check if the device has been interacting with this app if it's state when it started
+                // and now is the same
+                // Should be done in the background
+                checkIfDeviceKeyHasChanged(basicDeviceDetails, endpointId);
+            }
+
+        } else {
+            Timber.e(view.getString(R.string.log_hash_key_was_sent_in_an_invalid_format));
+            disconnectAndReset(endpointId);
+        }
+    }
+
+    @Override
+    public void sendLastReceivedRecords(@NonNull List<P2pReceivedHistory> receivedHistory) {
+        if (currentSender != null) {
+            interactor.sendMessage(new Gson().toJson(receivedHistory));
+            connectionLevel = ConnectionLevel.SENT_RECEIVED_HISTORY;
+        }
+    }
+
+    private void registerSendingDevice(Map<String, Object> basicDeviceDetails) {
+        SendingDevice sendingDevice = new SendingDevice();
+        sendingDevice.setDeviceId((String) basicDeviceDetails.get(Constants.BasicDeviceDetails.KEY_DEVICE_ID));
+        sendingDevice.setAppLifetimeKey((String) basicDeviceDetails.get(Constants.BasicDeviceDetails.KEY_APP_LIFETIME_KEY));
 
-                    Map<String, Object> map = (Map<String, Object>) new Gson()
-                            .fromJson(authenticationDetailsJson, Map.class);
+        P2PLibrary.getInstance().getDb()
+                .sendingDeviceDao()
+                .insert(sendingDevice);
+    }
+
+    @NonNull
+    private Integer clearDeviceHistoryAndUpdateDeviceKey(SendingDevice sendingDevice, String appLifetimeKey) {
+        AppDatabase db = P2PLibrary.getInstance().getDb();
+
+        sendingDevice.setAppLifetimeKey(appLifetimeKey);
+        db.sendingDeviceDao().update(sendingDevice);
+
+        return db.p2pReceivedHistoryDao()
```

Comment:
Shouldn't clear be done before update or maybe consider using [Transaction](https://developer.android.com/reference/android/arch/persistence/room/Transaction)

Reply:
Either can work, there is  `device_id` which will never change and is used to reference the device in the `received_history` table. I will add a transaction to this